File: src/handlers/update-workspace.go
package handlers

import (
	"fmt"
	"workspaces-service/src/models"
	"workspaces-service/src/utils"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

func UpdateWorkspace(c *fiber.Ctx) error {
	userID := fmt.Sprintf("%s", c.Locals("uuid"))

	workspaceID := c.Params("workspaceID", uuid.Nil.String())

	// parse body
	var body models.WorkspaceBody

	err := c.BodyParser(&body)
	if err != nil {
		fmt.Println(err.Error())
		return c.Status(400).JSON(fiber.Map{"message": "invalid-input"})
	}

	// check if workspace is not exist
	isExit, err := models.IsWorkspaceExist(workspaceID)

	if err != nil {
		return utils.ServerError(c, err)
	}

	if !isExit {
		return c.JSON(fiber.Map{"workspaceID": "This workspace does not exist"})
	}

	// check values
	if body.Workspace == "" {
		return c.JSON(fiber.Map{"workspace": "Must not be empty"})
	}

	// check if parent
	if body.ParentWorkspace != uuid.Nil {
		isExit, err := models.IsWorkspaceExist(body.ParentWorkspace.String())

		if err != nil {
			return utils.ServerError(c, err)
		}

		if !isExit {
			return c.JSON(fiber.Map{"parentWorkspace": "This workspace does not exist"})
		}
	}

	// update workspace
	err = models.UpdateWorkspace(body, workspaceID, userID)

	if err != nil {
		return utils.ServerError(c, err)
	}

	return c.JSON(fiber.Map{"message": "success"})
}

File: tests/routes/updateWorkspace.py
import utils
import config
from testRestApi import testRoute, Test, PUT

def testToken():
    res = testRoute(PUT, f"{config.server}/api/v1/workspaces/uuid")
    return res.equals({ "message": "invalid-token" })

def testEmptyFields():
    _, workspaceID = utils.createRandomWorkspace()
    res = testRoute(PUT, f"{config.server}/api/v1/workspaces/{workspaceID}", headers={ "X-Token": config.token })
    return res.equals({ "workspace": "Must not be empty" })

def testNotExistingParentWorkspace():
    randomWorkspace, workspaceID = utils.createRandomWorkspace()
    
    workspace = utils.randomString(10)
    parentWorkspace = utils.genUUID()
    body = { "workspace": workspace, "parentWorkspace": parentWorkspace }

    res = testRoute(PUT, f"{config.server}/api/v1/workspaces/{workspaceID}", headers={ "X-Token": config.token }, body=body)

    return res.equals({ "parentWorkspace": "This workspace does not exist"})

def testNotExistingWorkspace():    
    workspace = utils.randomString(10)
    workspaceID = utils.genUUID()
    body = { "workspace": workspace, "parentWorkspace": None }

    res = testRoute(PUT, f"{config.server}/api/v1/workspaces/{workspaceID}", headers={ "X-Token": config.token }, body=body)

    return res.equals({ "workspaceID": "This workspace does not exist"})

def testUpdateWorkspace():
    _, parentWorkspaceID = utils.createRandomWorkspace()
    _, workspaceID = utils.createRandomWorkspace()

    workspace = utils.randomString(10)
    body = { "workspace": workspace, "parentWorkspace": parentWorkspaceID }

    res = testRoute(PUT, f"{config.server}/api/v1/workspaces/{workspaceID}", headers={ "X-Token": config.token }, body=body)

    return res.equals({ "message": "success"})


tests = [
    Test("Update Workspace Invalid Token", testToken),
    Test("Update Workspace Empty Fields", testEmptyFields),
    Test("Update Workspace Not Existing Workspace", testNotExistingWorkspace),
    Test("Update Workspace Not Existing Parent Workspace", testNotExistingParentWorkspace),
    Test("Update Workspace", testUpdateWorkspace),
]
